llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp

```diff
@@ -31,10 +31,16 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST) {
   const LLT s32 = LLT::scalar(32);
   const LLT s64 = LLT::scalar(64);
 
+  const LLT nxv1s8 = LLT::scalable_vector(1, s8);
+  const LLT nxv2s8 = LLT::scalable_vector(2, s8);
+  const LLT nxv4s8 = LLT::scalable_vector(4, s8);
+  const LLT nxv8s8 = LLT::scalable_vector(8, s8);
+
+
   using namespace TargetOpcode;
 
   getActionDefinitionsBuilder({G_ADD, G_SUB, G_AND, G_OR, G_XOR})
-      .legalFor({s32, sXLen})
+      .legalFor({s32, sXLen, nxv1s8, nxv2s8, nxv4s8, nxv8s8})
```

Comment:
This needs to check ST.hasVInstructions()

Reply:
Apologies. RISC-V  V does not need predicates? I stated above SVE needs additional predicates.

Reply:
I think G_ADD has no predicates (all lanes are active) and we can use vadd.vv with or without a mask register. In order to implement G_ADD which has no predicates, we don't need need to generate a vadd.vv with a mask.

Reply:
> This needs to check ST.hasVInstructions()

Thanks for marking it out!

Sorry I don't quite follow the discussion above. Can someone explain the concept of predicate registers? I understand the masking and tail policy in riscv-v, and
1. what's the relation between mask/tail, and predicates?
2. how is predicate related to `ST.hasVInstructions()`?

Reply:
The word "predicate" has multiple meanings. The "has*" functions on RISCVSubtarget are known as predicates because they are used by `let Predicates = [HasVInstructions]` in tablegen.

Predicate also has meaning in the RISC-V vector spec to refer to using mask registers or VL to limit which elements are processed.

The two uses aren't connected.

Reply:
I see.
So I need to check `ST.hasVInstructions()` anyways?

Reply:
If we don't check hasVInstructions then we will legalize vector operations without having the ability to generate any RISC-V vector instructions. We need to check.